packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,268 @@
+        uint256 wPowerPerpToRedeem =
+            crabToRedeem.wmul(wPowerPerpInCrab).wdiv(IERC20(crab).totalSupply());
+
+        _zenBullSupply = _zenBullSupply.sub(_zenBullAmount);
+
+        IERC20(zenBull).transferFrom(msg.sender, address(this), _zenBullAmount);
+        _mint(msg.sender, _zenBullAmount);
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            3000,
+            wPowerPerpToRedeem,
+            _maxEthForWPowerPerp,
+            uint8(FLASH_SOURCE.EMERGENCY_WITHDRAW_ETH_FROM_CRAB),
+            abi.encodePacked(crabToRedeem, wPowerPerpToRedeem)
+        );
+
+        uint256 payout = IERC20(weth).balanceOf(address(this));
+        IWETH9(weth).withdraw(payout);
+        payable(msg.sender).sendValue(payout);
+
+        emit Withdraw(
+            msg.sender, _zenBullAmount, crabToRedeem, wPowerPerpToRedeem, payout, _zenBullAmount
+        );
+    }
+
+    /**
+     * @notice repay a portion of Euler debt and withdraw WETH collateral to this contract based on ZenBullEulerRecovery amount
+     * @dev this will revert of WETH to withdraw from Euler is greater than max, or limit price breach the tolerance %
+     * @dev if all ZenBullEulerRecovery are burnt, the ETH withdrawl will be activated
+     * @param _recoveryTokenAmount amount of ZenBullEulerRecovery token amount to burn
+     * @param _maxEthForUsdc max ETH to pay for 1 USDC
+     * @param _poolFee ETH/USDC Uni v3 pool fee
+     */
+    function emergencyRepayEulerDebt(
+        uint256 _recoveryTokenAmount,
+        uint256 _maxEthForUsdc,
+        uint24 _poolFee
+    ) external {
+        uint256 recoverShare = _recoveryTokenAmount.wdiv(totalSupply());
+        uint256 usdcToRepay = recoverShare.wmul(IEulerDToken(dToken).balanceOf(zenBull));
+        uint256 wethToWithdraw =
+            recoverShare.wmul(IEulerEToken(eToken).balanceOfUnderlying(zenBull));
+
+        require(
+            wethToWithdraw <= MAX_WETH_PER_DEBT_REPAY,
+            "WETH to withdraw is greater than max per repay"
+        );
+
+        uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, 420, false);
+
+        require(
+            _maxEthForUsdc <= ethUsdPrice.wmul((ONE.add(LIMIT_PRICE_TOLERANCE))),
+            "Max ETH for USDC greater than max price tolerance"
+        );
+
+        ethDividend[msg.sender] = ethDividend[msg.sender].add(_recoveryTokenAmount);
+        totalDividend = totalDividend.add(_recoveryTokenAmount);
+        _burn(msg.sender, _recoveryTokenAmount);
+
+        _exactOutFlashSwap(
+            weth,
+            usdc,
+            _poolFee,
+            usdcToRepay,
+            usdcToRepay.wmul(_maxEthForUsdc.mul(1e12)),
+            uint8(FLASH_SOURCE.EMERGENCY_REPAY_EULER_DEBT),
+            abi.encodePacked(usdcToRepay, wethToWithdraw)
+        );
+
+        IWETH9(weth).withdraw(IERC20(weth).balanceOf(address(this)));
+
+        if (
+            (IEulerEToken(eToken).balanceOfUnderlying(zenBull) == 0)
```

Comment:
is it possible that interest is accrued and we have like 2 wei left or something in contract

we only touch the protocol when we repay/withdraw, is it possible that balanceOf and balanceOfUnderlying() changes between initial call and after the repay

Reply:
still concerned about this, can we look at euler and see if this is a problem and also try to test maybe using past blocks or something?